metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanRetentionService.java

```diff
@@ -0,0 +1,193 @@
+package com.linkedin.metadata.entity.ebean;
+
+import com.linkedin.common.urn.Urn;
+import com.linkedin.metadata.dao.utils.RecordUtils;
+import com.linkedin.metadata.entity.EntityService;
+import com.linkedin.metadata.entity.RetentionService;
+import com.linkedin.retention.DataHubRetentionInfo;
+import com.linkedin.retention.Retention;
+import com.linkedin.retention.TimeBasedRetention;
+import com.linkedin.retention.VersionBasedRetention;
+import io.ebean.EbeanServer;
+import io.ebean.ExpressionList;
+import io.ebean.PagedList;
+import io.ebean.Transaction;
+import io.opentelemetry.extension.annotations.WithSpan;
+import java.sql.Timestamp;
+import java.time.Clock;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import static com.linkedin.metadata.Constants.ASPECT_LATEST_VERSION;
+
+
+@Slf4j
+@RequiredArgsConstructor
+public class EbeanRetentionService extends RetentionService {
+  private final EntityService _entityService;
+  private final EbeanServer _server;
+  private final int _batchSize;
+
+  private final Clock _clock = Clock.systemUTC();
+
+  public EntityService getEntityService() {
+    return _entityService;
+  }
+
+  @Override
+  @WithSpan
+  public void applyRetention(@Nonnull Urn urn, @Nonnull String aspectName, List<Retention> retentionPolicies,
+      Optional<RetentionService.RetentionContext> retentionContext) {
+    log.debug("Applying retention to urn {}, aspectName {}", urn, aspectName);
+    // If no policies are set or has indefinite policy set, do not apply any retention
+    if (retentionPolicies.isEmpty() || retentionPolicies.get(0).hasIndefinite()) {
+      return;
+    }
+    ExpressionList<EbeanAspectV2> deleteQuery = _server.find(EbeanAspectV2.class)
+        .where()
+        .eq(EbeanAspectV2.URN_COLUMN, urn.toString())
+        .eq(EbeanAspectV2.ASPECT_COLUMN, aspectName)
+        .ne(EbeanAspectV2.VERSION_COLUMN, ASPECT_LATEST_VERSION)
+        .or();
+    boolean retentionApplied = false;
+    for (Retention retention : retentionPolicies) {
+      if (retention.hasVersion() && applyVersionBasedRetention(urn, aspectName, deleteQuery, retention.getVersion(),
```

Comment:
This code initially confused me because I didn't realize that deleteQuery was being mutated inside of applyVersionBasedRetention as "querySoFar"... what if that method simply returned a query and we inserted it into the expression list here?

Reply:
That was what I wanted to do, but didn't think there was a way. Looking again, there might be a way. let me see